### ISSUE TITLE
Model compilation support

### ISSUE THREAD

**Issue by ipoletaev** [Lightning-AI/lit-llama#62]
With FSDP currently the code could not be run.

If you try to add model compilation to the [training](https://github.com/Lightning-AI/lit-llama/blob/main/train.py) like:
```
...
fabric = L.Fabric(accelerator="cuda", devices=8, precision="bf16-mixed", strategy=strategy)
fabric.launch()
...

model = fabric.setup_module(model)
# compile() goes should go wrapping as per https://github.com/huggingface/transformers/commit/fb0a38b4f275727d6228fb4a78c15c6dd8480e91
# Though it does not work either even if goes before setup_module() as you'll get
model = torch.compile(model)

optimizer = torch.optim.AdamW(model.parameters(), ...)
optimizer = fabric.setup_optimizers(optimizer)
...

train(model, ...)
```

and try it via:

```
lightning run model --accelerator=cuda --devices=8 train.py ...
```

you'll get:

```
  File ".../.venv/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 172, in __call__
    return self._wrap(value).clone(**self.options())
  File ".../.venv/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 345, in _wrap
    assert getattr(
AssertionError: Dynamo only supports FSDP with use_orig_params=True
```

If I pass `use_orig_params = True` into the `FSDPStrategy()` constructor you get:

```
ValueError: The optimizer does not seem to reference any FSDP parameters. HINT: Make sure to create the optimizer after setting up the model.
```

So, then one can remove `optimizer = fabric.setup_optimizers(optimizer)` as it is anyway no-op for FSDP, but even in this case I see:

```
from user code:
   File ".../.venv/lib/python3.8/site-packages/lightning_utilities/core/apply_func.py", line 75, in apply_to_collection
    is_namedtuple_ = is_namedtuple(data)
```
---

Also I was not able to find any tests of compiled model with the FSDP neither [here](https://github.com/Lightning-AI/lightning/blob/master/tests/tests_fabric/strategies/test_fsdp.py) nor [here](https://github.com/Lightning-AI/lightning/blob/master/tests/tests_fabric/strategies/test_fsdp_integration.py).

I wonder if anyone was able to successfully launch compiled model in a FSDP regime? Thanks a lot for the help!

**Comment by ipoletaev** [comment]
Ok, looks like figured a few things out. One needs to:

0) Disable activations checkpointing
1) Pass `use_orig_params=True` to the strategy
2) Compile model [inside fabric setup_module()](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/fabric.py#L246) like this:

```
...
module = self._strategy.setup_module(module)
module = torch.compile(module)
module = _FabricModule(module, self._precision, original_module=original_module)
...
```

3) Remove [this check](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/strategies/fsdp.py#L334) on the model type for sync()

---

TL;DR: one has to compile `FullyShardedDataParallel()`-wrapped model and use the result object to create the optimizer.

**Comment by lantiga** [comment]
Great find @ipoletaev

**Comment by ipoletaev** [comment]
Yep. This requires quite a bit of refactoring in the lightning fabric code. 

E.g. we'd need to add optional bool argument to compile the model inside `model_setup()` and change model wrapping logic a bit as in FSDP for evals you need to run things on top of the original model:

```
_FabricModule( ### CPU!!!!
  (_forward_module): OptimizedModule( ### GPU
    (_orig_mod): FullyShardedDataParallel( ### GPU
      (module): your_model_class() ....
```
i.e. you need `model._forward_module._orig_mod` and not `model._forward_module` as `_FabricModule` does it now to run evals.

**Comment by awaelchli** [comment]
Hey @ipoletaev So the setup logic in Fabric was produced at a time when FSDP didn't have the use_orig_params yet. This is a new feature. But the plan was to standardize the setup for FSDP the same way as it is for the other strategies, namely:

```py
model = ...
optimizer = Adam(model.parameters())
model, optimizer = fabric.setup(model, optimizer)
```
I just haven't gotten around doing that.
This way, the user doesn't have to change this part of the code when switching strategies. This is one of the improvements I am looking forward to for FSDP in Fabric. This should then also make it possible to `torch.compile` after the setup, and PyTorch itself is also working on improving the support for compiling an FSDP wrapped module.

**Comment by ipoletaev** [comment]
Great!

It is just the trick one has to do with FSDP specifically during eval steps (e.g. even before calling `model.eval()`) to call `loss = model(**data)` the model has to be a type of `FullyShardedDataParallel` vs `OptimizedModule` because the former is what is on GPUs as mentioned [here](https://github.com/Lightning-AI/lit-llama/issues/62#issuecomment-1492387824).

If you just do `model(**data)` as usually with FSDP I am getting errors related to the fact that the model (the one that comes out of `fabric.setup_module`) is on CPU.

**Comment by ipoletaev** [comment]
Also we need to figure out the magic required to enabled activation (aka gradient checkpointing) for the compiled FSDP model. This is actually the "biggest deal" for LLaMA to be able to run really big models.

**Comment by awaelchli** [comment]
I just linked two issues on the Lighting/Fabric side. In the mean time, I think we should at least try to make the LLama model torch.compilable on its own, that's the first step (#21). 

Agreed. I don't believe that FSDP/torch.compile can support activation checkpointing atm, we should check related issues in the pytorch side. 

FYI: Since you mentioned FakeTensorMode, we also discussed supporting that in the future but PyTorch still has some unimplemented parts there (https://github.com/Lightning-AI/lightning/issues/16448)

**Comment by ipoletaev** [comment]
Here is what made things work for me (please feel free to adapt this): 

1) Patch lightning's FSDP setup module:

```
    def setup_module(self, module: Module, compile: bool = False) -> "FullyShardedDataParallel":
        """Wraps the model into a
        :class:`~torch.distributed.fsdp.fully_sharded_data_parallel.FullyShardedDataParallel` module."""
        from torch.distributed.fsdp.fully_sharded_data_parallel import FullyShardedDataParallel

        if "auto_wrap_policy" in self._ddp_kwargs and any(
            isinstance(mod, FullyShardedDataParallel) for mod in module.modules()
        ):
            # If model is already wrapped, we need to avoid sending the `auto_wrap_policy`
            del self._ddp_kwargs["auto_wrap_policy"]

        wrapped_module = FullyShardedDataParallel(
            module=module,
            cpu_offload=self.cpu_offload,
            mixed_precision=self.mixed_precision_config,
            device_id=self.root_device.index,
            **self._ddp_kwargs,
        )

        # activation checkpointing needs to be set up after wrapping the model
        if _TORCH_GREATER_EQUAL_1_13 and self._activation_checkpointing:
            _setup_activation_checkpointing(module=wrapped_module, layers=self._activation_checkpointing)
            
        if compile:
            wrapped_module = torch.compile(wrapped_module)

        return wrapped_module
 ```
 
 2) create strategy like this:
 
 ```
 strategy = fsdp_strategy.FSDPStrategy(
                auto_wrap_policy=functools.partial(
                    wrap.transformer_auto_wrap_policy,
                    transformer_layer_cls={Block}),
                activation_checkpointing=None
                cpu_offload=True,
                use_orig_params=True)
```

3) create model like this:


```
model.train()
model = fabric.setup_module(model, compile=true)
optimizer = deepspeed.ops.adam.DeepSpeedCPUAdam(model._forward_module.parameters(), ...)
... training loop....
```

**Comment by ipoletaev** [comment]
For ones who still wants to merry model compilation and FSDP training. There are a few more problems you are about to face:

1) Checkpointing model's state for further serving. It is just super complex API like:

```
model = ... # FullShardedDataParallel type
state_dict_config = fsdp.FullStateDictConfig(offload_to_cpu=True, rank0_only=True)
fsdp_state_dict_type = fsdp.StateDictType.FULL_STATE_DICT
with fsdp.FullyShardedDataParallel.state_dict_type(
            model,
            state_dict_type=fsdp_state_dict_type,
            state_dict_config=state_dict_config):
    model_state = model.state_dict()

torch.save(model_state,  ...)
```

2) Checkpointing optimizer's state to resume training if needed:

```
model = ... # FullShardedDataParallel type
optimizer_state = fsdp.FullyShardedDataParallel.full_optim_state_dict(model=model, optim=optimizer, rank0_only=True)
torch.save(optimizer_state,  ...)
```

does not work with 

```
  File "../torch/distributed/fsdp/fully_sharded_data_parallel.py", line 1330, in full_optim_state_dict
    return FullyShardedDataParallel._optim_state_dict_impl(
  File "../torch/distributed/fsdp/fully_sharded_data_parallel.py", line 1154, in _optim_state_dict_impl
    return _optim_state_dict(
  File "../torch/distributed/fsdp/_optim_utils.py", line 1463, in _optim_state_dict
    unflat_state = _unflatten_optim_state(
  File "../torch/distributed/fsdp/_optim_utils.py", line 136, in _unflatten_optim_state
    consolidated_state = _communicate_optim_state(
  File "../torch/distributed/fsdp/_optim_utils.py", line 209, in _communicate_optim_state
    dist.all_gather_into_tensor(
  File "../torch/distributed/distributed_c10d.py", line 1436, in wrapper
    return func(*args, **kwargs)
  File "../torch/distributed/distributed_c10d.py", line 2517, in all_gather_into_tensor
    work = group._allgather_base(output_tensor, input_tensor)
RuntimeError: output tensor size must be equal to world_size times input tensor size
```

**Comment by awaelchli** [comment]
@ipoletaev There is some work here: https://github.com/Lightning-AI/lightning/pull/17323
Just a note on your code snippets above, if you do `rank0_only=True` you should also make sure you only save on rank 0.

**Comment by ipoletaev** [comment]
That's right (though other processes just get empty dicts). I corrected the snippet.

**Comment by awaelchli** [comment]
Fabric doesn't do this, it is the user's responsibility at the moment. Capturing and restoring the random state inside the dataloaders workers is non-trivial. This degree of fault-tolerance was tested in Lightning Trainer but was never successful. If you have feedback and suggestions on this, feel free to post an issue on https://github.com/Lightning-AI/lightning/issues  with your proposal. 

Let's keep the topic of this issue on the model compilation.

**Comment by ipoletaev** [comment]
[Related](https://discuss.pytorch.org/t/gpu-memory-spikes-with-fsdp-and-torch-compile/179088/4).
